### PR TITLE
streaming support

### PR DESCRIPTION
**Pull request by rocketbitz**
Added go client support, but still need to add python client support. I will add a small write-up of the protocol in the readme, and plan to add an example once the python client code is written.

### CODE REVIEW

**Comment by rocketbitz** [comment]
Also need to handle higher timeframes, and detecting when the window is closed.

**Comment by rocketbitz** [comment]
With the implementation of the expiration shelf on the streaming interface in my last commit, the aggregate streaming should be handled in the best way it can be. Aggregated data entries will be placed on a 'shelf' and will be given a shelf-life that indicates when they will be delivered. These entries will be updated with the latest aggregated candle up until the expiration time. For now, that expiration time is the upper bound of the aggregate timeframe, but a buffer duration will likely be necessary to account for any delays in writing the final 1Min candle. I may make this configurable, but in the effort of avoiding unnecessary configuration flags, I am going to do some live testing first before deciding whether or not it's required.

File: frontend/stream/stream.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+	streams map[string]struct{}
+}
+
+// Subscribed matches the subscriber's subscribed streams
+// with the supplied timebucket key string.
+func (s *Subscriber) Subscribed(itemKey string) bool {
+	s.RLock()
+	defer s.RUnlock()
+	for stream := range s.streams {
+		g := glob.MustCompile(stream, '/')
```
Comment (umitanuki) [inline frontend/stream/stream.go]: Should be `Compile`.  You don't want to panic by malicious input

**Comment by blonc** [comment]
if i am correct, this would allow you to retrieve data in real time from what ever the source is, what I am not clear on is, is the data pulling it from the database or is it streaming directly from the source?

**Comment by umitanuki** [comment]
You are right, the intention of this is to push the new data to the client immediately.  We are trying to achieve it by using the trigger mechanism and currently triggers can pull data from disk only.  We could add a parameter for new data for trigger that's just written on the disk, but in many cases trigger anyway needs to check the data on disk (in this case, read the base timeframe bar to know if the lower resolution candle is ready or not)

File: frontend/stream/stream.go (hunk 1 above)
Comment (rocketbitz) [reply frontend/stream/stream.go]: It's safely compiled on subscribe here: https://github.com/alpacahq/marketstore/pull/58/files#diff-9c793816e90787e8eb510acd93ab89a4R128. Once it's in memory, it's never modified, only replaced by new subscribe messages, which are also safely validated, so this is safe.

**Comment by blonc** [comment]
how long till you think the python client is wrote? I am going to be implementing this system into a framework called backtrader and obviously having the creator write the code will work much better than me trying to take gocode and push the results to a python framework.  Is there any financial support I can provide to speed it up?

**Comment by blonc** [comment]
I sent you a message there

________________________________

File: contrib/stream/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+```
+$ make all
+```
+
+It installs the new .so file to the first GOPATH/bin directory.
+
+
+## Caveat
+Since this is implemented based on the Go's plugin mechanism, it is supported only
```
Comment (umitanuki) [inline contrib/stream/README.md]: 1.10 support Mac as well?

Comment (rocketbitz) [reply contrib/stream/README.md]: Oh ya, didn't realize they added support for MacOS. I will update.

File: frontend/stream/stream.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+				return fmt.Errorf("%s is an invalid stream", stream)
+			}
+			m[stream] = struct{}{}
+		}
+		s.streams = m
+	}
+	return nil
+}
+
+func validStream(stream string) bool {
```
Comment (umitanuki) [inline frontend/stream/stream.go]: You should consult catalog / TimeBucketKey

Comment (rocketbitz) [reply frontend/stream/stream.go]: Catalog will not be able to match with * notation.

File: frontend/stream/stream.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
+package stream
```
[the remark this answers is not in the document]
Comment (rocketbitz) [reply frontend/stream/stream.go]: I improved readme. We have other priorities now, so will merge this as is.

File: frontend/stream/stream.go (hunk 3 above)
Comment (umitanuki) [reply frontend/stream/stream.go]: Right, but I think we would like to accept non-existent symbol that are put by client by accident

File: frontend/stream/stream.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+			}
+		}
+
+		catalog.RUnlock()
+	}
+}
+
+// Payload is used to send data over the websocket
+type Payload struct {
+	Key  string
```
Comment (umitanuki) [inline frontend/stream/stream.go]: Missing field tag. I think we should do lower case field names to be consistent with rest of the system.

**Pull request merged**